sway/desktop/transaction.c

```diff
@@ -0,0 +1,398 @@
+	}
+	if (view->surface && wlr_surface_has_buffer(view->surface)) {
+		wlr_buffer_ref(view->surface->buffer);
+		instruction->saved_buffer = view->surface->buffer;
+	}
+}
+
+static void transaction_destroy(struct sway_transaction *transaction) {
+	// Free instructions
+	for (int i = 0; i < transaction->instructions->length; ++i) {
+		struct sway_transaction_instruction *instruction =
+			transaction->instructions->items[i];
+		struct sway_container *con = instruction->container;
+		for (int j = 0; j < con->instructions->length; ++j) {
+			if (con->instructions->items[j] == instruction) {
+				list_del(con->instructions, j);
+				break;
+			}
+		}
+		if (con->destroying && !con->instructions->length) {
+			container_free(con);
+		}
+		remove_saved_view_buffer(instruction);
+		free(instruction);
+	}
+	list_free(transaction->instructions);
+
+	// Free damage
+	list_foreach(transaction->damage, free);
+	list_free(transaction->damage);
+
+	if (transaction->timer) {
+		wl_event_source_remove(transaction->timer);
+	}
+	free(transaction);
+}
+
+static void copy_pending_state(struct sway_container *container,
+		struct sway_container_state *state) {
+	state->layout = container->layout;
+	state->swayc_x = container->x;
+	state->swayc_y = container->y;
+	state->swayc_width = container->width;
+	state->swayc_height = container->height;
+	state->has_gaps = container->has_gaps;
+	state->current_gaps = container->current_gaps;
+	state->gaps_inner = container->gaps_inner;
+	state->gaps_outer = container->gaps_outer;
+	state->parent = container->parent;
+
+	if (container->type == C_VIEW) {
+		struct sway_view *view = container->sway_view;
+		state->view_x = view->x;
+		state->view_y = view->y;
+		state->view_width = view->width;
+		state->view_height = view->height;
+		state->is_fullscreen = view->is_fullscreen;
+		state->border = view->border;
+		state->border_thickness = view->border_thickness;
+		state->border_top = view->border_top;
+		state->border_left = view->border_left;
+		state->border_right = view->border_right;
+		state->border_bottom = view->border_bottom;
+	} else if (container->type == C_WORKSPACE) {
+		state->ws_fullscreen = container->sway_workspace->fullscreen;
+		state->ws_floating = container->sway_workspace->floating;
+		state->children = create_list();
+		list_cat(state->children, container->children);
+	} else {
+		state->children = create_list();
+		list_cat(state->children, container->children);
+	}
+}
+
+static bool transaction_has_container(struct sway_transaction *transaction,
+		struct sway_container *container) {
+	for (int i = 0; i < transaction->instructions->length; ++i) {
+		struct sway_transaction_instruction *instruction =
+			transaction->instructions->items[i];
+		if (instruction->container == container) {
+			return true;
+		}
+	}
+	return false;
+}
+
+void transaction_add_container(struct sway_transaction *transaction,
+		struct sway_container *container) {
+	if (transaction_has_container(transaction, container)) {
```

Comment:
What if the container has already been added to the transaction, but its state has changed since last time?

Reply:
Theoretically it would configure and apply the first state which is bad, but this doesn't happen in practice. To make this happen you would have to make a transaction the manual way, add the container to it, modify something in the container, then add the container again. You should always use do them in the order: modify tree, add to transaction, commit transaction.

Reply:
Done.